### ISSUE TITLE
plugins.youtube Receiving "Unable to open URL: https://youtube.com/get_video_info" when trying to do --retry-streams after a few retry attempts.

### ISSUE THREAD

**Issue by xanek** [streamlink/streamlink#3724]
## Plugin Issue

- [x] This is a plugin issue and I have read the contribution guidelines.
- [x] I am using the latest development version from the master branch.

### Description

When trying to use the --retry-streams options on a YouTube livestream that has not begun yet, shortly after running it with either 1-20 seconds between when it checks, it will start outputting

`[cli][error] Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=yZ7CWjuMCe8&eurl=https%3A%2F%2Fyoutube.googleapis.com%2Fv%2FyZ7CWjuMCe8)`

It used to not do this at all until recently about 2-3 days ago on streamlink 2.0.0 and when it did, I tried updating to the current development build but it is still giving me the same thing.

So I am unable to setup streamlink to record for a stream that I wouldn't be around for since it would give that error even after the stream would start.

### Reproduction steps / Explicit stream URLs to test

1. input `streamlink --retry-streams <#> -o <name.ts> <link of a stream that is planned to go live later>`

### Log output

```
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.8.7
[cli][debug] Streamlink: 2.1.1+14.g2884485
[cli][debug] Requests(2.25.1), Socks(1.7.1), Websocket(0.59.0)
[cli][debug] Arguments:
[cli][debug]  --loglevel=debug
[cli][debug]  --rtmpdump=C:\Program Files (x86)\Streamlink\rtmpdump\rtmpdump.exe
[cli][debug]  --ffmpeg-ffmpeg=C:\Program Files (x86)\Streamlink\ffmpeg\ffmpeg.exe
```

**Comment by llyyr** [comment]
This appears to be a Youtube issue, you can get a 404 on get_video_info if you try to access it too quickly on any URL, not just planned streams or live streams. It appears to be fine if you wait 1 second between requests to get_video_info though.

**Comment by xanek** [comment]
> This appears to be a Youtube issue, you can get a 404 on get_video_info if you try to access it too quickly on any URL, not just planned streams or live streams. It appears to be fine if you wait 1 second between requests to get_video_info though.

I've tried 2 seconds and 15 seconds and it still does that. It seems to not do it when I try for 30 seconds, but otherwise it doesn't seem to be possible to do anywhere from 1-15 or 20 seconds for the past 4 days.

**Comment by jenlunds** [comment]
I use 5 minutes with --retry-streams and I still get the error sometimes. I've also gotten it when the stream is already live and I've included --retry-streams in the command. This is the only request I make to youtube at that time, its not like I have a lot of streams waiting to go live when the error appears. It used to work fine before so I don't know what happened.

**Comment by scottbernstein** [comment]
There is some kind of YouTube DDOS protection that kicks in sometimes. There is nothing you can do to stop it. It is on the YouTube server side. All you can do is wait a few days until your IP address is unblocked.  It will clear eventually on its own.  I have experienced this as well.

Sent from my iPhone

On May 16, 2021, at 1:07 PM, jenlunds ***@***.***> wrote:

﻿

I use 5 minutes with --retry-streams and I still get the error sometimes. I've also gotten it when the stream is already live and I've included --retry-streams in the command. This is the only request I make to youtube at that time, its not like I have a lot of streams waiting to go live when the error appears. It used to work fine before so I don't know what happened.

**Comment by Kethsar** [comment]
It's feels too inconsistent to be some kind of DDoS protection.
Seems more like some of their servers are a little screwy from some update that might have been pushed around a week ago. It's affected the RSS of someone I know as well, giving him a lot of 404s recently.

@xanek, does it happen for you 100% of the time, most of the time, occasionally, etc.?
I hit it surprisingly often with streamlink compared to one of my own scripts, and when I was testing with curl. More interesting is that if it gets hit, it continues to be hit for that running instance of streamlink. Stopping and immediately restarting will usually work, even if there's <2 seconds between the failed request of the old instance, and the request from the new instance. I imagine this has to do with requests keeping the connection open, and thus whatever server the youtube proxies forwarded you to continues to be used, so if it is a bad one, it just will not work. I've set `--retry-streams 60` for one test to see if that would be long enough for requests to make a completely new connection, but it seems not.

I've opted for pulling the player_response object from a `script` tag in the watch page as a fallback for my script, disgusting as that is. It's simple enough though, as the script tag that declares it is used solely for the declaration, so there little that needs to be stripped before using it. For now.

**Comment by xanek** [comment]
> @xanek, does it happen for you 100% of the time, most of the time, occasionally, etc.?
Ever since a few days ago, it happens all the time, never was a problem for months.

**Comment by xanek** [comment]
> @xanek, does it happen for you 100% of the time, most of the time, occasionally, etc.?

This has been happening ever since a few days ago.
It never happened prior to that and I've been using streamlink for more than half a year already.

**Comment by scottbernstein** [comment]
This has the same characteristics of what I have experienced. Like I said, it’s definitely some kind of DDOS protection against this URL. Your IP is on a banned list. It will go away on its own after 3/4/5 days or something. There is nothing at all you can do about it as it’s on the server side.

You can tell it’s not a streamlink issue because when this is occurring, the youtube-dl application also exhibits the same issue.

Sent from my iPhone

On May 16, 2021, at 9:26 PM, xanek ***@***.***> wrote:

﻿

@xanek<https://github.com/xanek>, does it happen for you 100% of the time, most of the time, occasionally, etc.?

This has been happening ever since a few days ago.
It never happened prior to that and I've been using streamlink for more than half a year already.

**Comment by xanek** [comment]
> This has the same characteristics of what I have experienced. Like I said, it’s definitely some kind of DDOS protection against this URL. Your IP is on a banned list. It will go away on its own after 3/4/5 days or something. There is nothing at all you can do about it as it’s on the server side. You can tell it’s not a streamlink issue because when this is occurring, the youtube-dl application also exhibits the same issue.

Wouldn't it completely lock me out from using streamlink in general then?
I am able to still use it, just not able to have it retry constantly.

**Comment by scottbernstein** [comment]
It is not necessarily 100% consistent…so yes this makes sense.

Sent from my iPhone

On May 16, 2021, at 9:31 PM, xanek ***@***.***> wrote:

﻿

This has the same characteristics of what I have experienced. Like I said, it’s definitely some kind of DDOS protection against this URL. Your IP is on a banned list. It will go away on its own after 3/4/5 days or something. There is nothing at all you can do about it as it’s on the server side. You can tell it’s not a streamlink issue because when this is occurring, the youtube-dl application also exhibits the same issue.

Wouldn't it completely lock me out from using streamlink in general then?
I am able to still use it, just not able to have it retry constantly.

**Comment by Kethsar** [comment]
> It is not necessarily 100% consistent

That would make for some pointless DDoS protection, I must say. Not to mention, at least in my case, it started happening after using streamlink for the first time in almost a week, and I was not exactly spamming downloads with other tools in that time either. 
You are correct it's not a streamlink issue though, at least it's not the cause. But if it ends up that this stays an issue for a long while, streamlink would probably want to try to figure out some way to mitigate it.

**Comment by scottbernstein** [comment]
Well the logic of their DDOS protection may indeed be faulty, but unless you are assigned a static IP address from your ISP you have no idea what the person who has your IP before you did was doing.

I have had this happen to me on at least 3 occasions. There is to way to tell what triggered it and there is no way to get around it, except to wait for it to clear out. Trust me.

Sent from my iPhone

On May 16, 2021, at 9:52 PM, Kethsar ***@***.***> wrote:

﻿

It is not necessarily 100% consistent

That would make for some pointless DDoS protection, I must say. Not to mention, at least in my case, it started happening after using streamlink for the first time in almost a week, and I was not exactly spamming downloads with other tools in that time either.
You are correct it's not a streamlink issue though, at least it's not the cause. But if it ends up that this stays an issue for a long while, streamlink would probably want to try to figure out some way to mitigate it.

**Comment by xanek** [comment]
Well I remember the last time I had something similar to where I couldn't even use streamlink at all because of being banned, that I had to use a VPN to get around it.

But even with a VPN right now it is giving me the same or similar messages.

**Comment by xanek** [comment]
> It is not necessarily 100% consistent…so yes this makes sense.

Well I remember the last time I had something similar to where I couldn't even use streamlink at all because of being banned, that I had to use a VPN to get around it.

But even with a VPN right now it is giving me the same message.

**Comment by tryoutuser** [comment]
is it really DDoS? i tried with youtube-dl and it parse different list of m3u8 contents for an ongoing livestream video.

it's no longer listing webm, mp4 or audio and video only formats.

![image](https://user-images.githubusercontent.com/35561383/118475579-a0deb900-b736-11eb-9be0-fd4a42d028e9.png)

**Comment by Kethsar** [comment]
That is a livestream you checked. Livestreams only have VP9 (webm) if they are using Low Latency mode. Ultra-Low Latency and Normal Latency do not have VP9 as an option. That stream is Normal latency.
Even if it were Low Latency mode, youtube-dl and I think streamlink both grab from the HLS stream playlists, which youtube only provides H264+AAC combined, and thus no audio-only or VP9.

**Comment by highflux7** [comment]
Youtube has again changed the way the stream info is presented.  Even the (slightly older) VLC v3.0.9.2 now barfs on most youtube urls.

**Comment by xanek** [comment]
So would we have to wait for a new update with streamlink or is this something that can't be workaround?

**Comment by vagosss** [comment]
Mpc-be also has issues with the built in parser for youtube... https://sourceforge.net/p/mpcbe/tickets/819/

**Comment by yavuzKomecoglu** [comment]
@xanek, Does [this PR](https://github.com/streamlink/streamlink/pull/3732) solve the problem?

**Comment by xanek** [comment]
> @xanek, Does [this PR](https://github.com/streamlink/streamlink/pull/3732) solve the problem?

I think it is? I'm running it on a stream that's not live yet with "--retry-streams 2" and it doesn't seem to be giving those messages anymore.

Just to make sure, I'm not sure how to get that pull request, but I went and modified the youtube.py with the changes that were noted and seems to be working fine as of now.

**Comment by yavuzKomecoglu** [comment]
We updated (v2.1.2) our production application. There is no problem.

**Comment by xanek** [comment]
> We updated (v2.1.2) our production application. There is no problem.

Seems like the problem is starting up again.

```
[cli][error] Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=bHJ_favp9MI&html5=1&eurl=https%3A%2F%2Fyoutube.googleapis.com%2Fv%2FbHJ_favp9MI)
[cli][error] Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=bHJ_favp9MI&html5=1&el=detailpage)
[cli][error] Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=bHJ_favp9MI&html5=1&el=detailpage)
```

I'm unsure what is causing it now.

**Comment by yavuzKomecoglu** [comment]
> > We updated (v2.1.2) our production application. There is no problem.
> 
> Seems like the problem is starting up again.
> 
> ```
> [cli][error] Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=bHJ_favp9MI&html5=1&eurl=https%3A%2F%2Fyoutube.googleapis.com%2Fv%2FbHJ_favp9MI)
> [cli][error] Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=bHJ_favp9MI&html5=1&el=detailpage)
> [cli][error] Unable to open URL: https://youtube.com/get_video_info (404 Client Error: Not Found for url: https://www.youtube.com/get_video_info?video_id=bHJ_favp9MI&html5=1&el=detailpage)
> ```
> 
> I'm unsure what is causing it now.

Thanks for your feedback. Our system is in a similar situation. I also re-examine and find a solution.

**Comment by bastimeyer** [comment]
Please open a new issue thread.
